core/sourcehook/sh_memory.h

```diff
@@ -58,9 +59,12 @@ namespace SourceHook
 #if SH_SYS == SH_SYS_LINUX
 		// On linux, first check /proc/self/maps
 		unsigned long laddr = reinterpret_cast<unsigned long>(addr);
-
 		bool bFound = false;
-		FILE *pF = fopen("/proc/self/maps", "r");
+
+		// Open once statically, duplicate before using to stay safe
+		// Way way faster than reopening from scratch
+		static std::unique_ptr<FILE, int(*)(FILE*)> pMasterFile(fopen("/proc/self/maps", "r"), fclose);
+		FILE *pF = fdopen(dup(fileno(pMasterFile.get())), "r");
```

Comment:
Isn't this not threadsafe? dup() doesn't give you separate seek offsets. It's basically just a refcount.

I'm curious why this is faster, anyway. Is it because it's skipping the VFS lookup? Or is fopen() doing something very expensive? Has anyone looked at a perf trace?

Reply:
Two alternatives that would be threadsafe.

One is to open() /proc/self, and keep it open. Then you can openat(), which probably skips some of the VFS overheard in the kernel.

Another is to see if dumping fopen() in favor of open()/read() is better, or if std::ifstream is better.

Finally, if you truly want to only ever open the file once... use C++11's "thread_local" keyword. This should be in a .cpp file, not in this inline function, so you'd want to make a helper function. Historically thread local slots were a precious resource. That might not be true anymore but better to be on the safe side.

Reply:
Yeah you are correct, I didn't test thoroughly enough so I didn't catch the offset being tracked. Thread_local does improve it a bit but I don't know if it's worth it noting your concerns above. Realistically a lot of the overhead that is left is probably due to the kernel using printf for the map output

Reply:
Maybe the next question to answer is "can we reduce the number of calls to GetPageBits?"

Metamod:Source is designed to exist in a closed system, eg, there are no other vtable hooks going on outside of SourceHook. And calling mprotect() on text pages is pretty unusual, so in theory, MM:S is the only place flipping those bits on and off.

Maybe MM:S can keep a global cache of which pages have been made writable, and then never make them readonly again.

Reply:
One thing that would help sourcemod specifically is removing these parts of sdkhooks: https://github.com/alliedmodders/sourcemod/blob/e88e9342efd4088ed4e560da988e0c7329c91151/extensions/sdkhooks/extension.cpp#L794C1-L799C5. I did a test where I deleted each occurrence of a hook being deleted which was a big speedup, but I am not sure whether that would mess anything up.